### ISSUE TITLE
copy_tree()'s .st_dev cross file system checks incomptible with overlayfs

### ISSUE THREAD

**Issue by xnox** [systemd/systemd#9134]
**systemd version the issue has been seen with**

v238+

(many prior versions too

**Used distribution**
Ubuntu

**Expected behaviour you didn't see**
test-copy passes

**Unexpected behavior you saw**
```
# ./test-copy
test_copy_file
test_copy_file_fd
test_copy_tree
Assertion 'access(f, F_OK) == 0' failed at ../src/test/test-copy.c:138, function test_copy_tree(). Aborting.
Aborted (core dumped)
```

The actual assert is that 'file' was not copied, and is not present in the target dir:

```
# tree /tmp/test-copy_tree*
/tmp/test-copy_tree
├── dir1
│   ├── dir2
│   │   ├── dir3
│   │   │   └── dir4
│   │   │       └── dir5
│   │   │           └── file
│   │   └── file
│   └── file
├── file
├── link -> /tmp/test-copy_tree/file
├── link2 -> /tmp/test-copy_tree/dir1/file
└── unixsock
/tmp/test-copy_tree-copy
└── dir1
    └── dir2
        └── dir3
            └── dir4
                └── dir5

10 directories, 7 files
```

**Steps to reproduce the problem**
setup overlayfs of `/tmp`, execute `./test-copy` binary
In my case, I use sbuild with overlayfs, thus I'm in a chroot at `/run/schroot/mount/cosmic-amd64-ae810e2c-3076-4d66-90de-b55c204850af` which is `cosmic-amd64 on /run/schroot/mount/cosmic-amd64-ae810e2c-3076-4d66-90de-b55c204850af type overlay (rw,relatime,lowerdir=/var/lib/schroot/union/underlay/cosmic-amd64-ae810e2c-3076-4d66-90de-b55c204850af,upperdir=/var/lib/schroot/union/overlay/cosmic-amd64-ae810e2c-3076-4d66-90de-b55c204850af/upper,workdir=/var/lib/schroot/union/overlay/cosmic-amd64-ae810e2c-3076-4d66-90de-b55c204850af/work)`

But a simple setup of `/tmp` as overlayfs should do the trick too, e.g.:
```
overlay on /tmp type overlay (rw,relatime,lowerdir=/run/schroot/mount/cosmic-amd64-ae810e2c-3076-4d66-90de-b55c204850af/mnt/lower,upperdir=/mnt/upper,workdir=/mnt/work)
```
or similar.

I'm not sure if this is related to https://github.com/systemd/systemd/issues/3847 or not.

**Comment by mbiebl** [comment]
See also
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=854400
and
https://github.com/systemd/systemd/pull/6656

**Comment by xnox** [comment]
Thanks @mbiebl.

Just to add, the bug here is not that /tmp is overlayfs (that is a bit silly), but that all of / is.

And it seems like a sort of thing one may do when launching ephemeral containers.
It seems like this would break systemd-tmpfiles, and thus booting overlayfs backed containers with e.g. nspawn. I will check if LXD offers overlayfs backend to container instances. And ping kernel engineering.

So far, my local workaround is to mount a tmpfs on top of /tmp, but that would not help with e.g. tmpfiles.d processing.

**Comment by poettering** [comment]
So, what is this about precisely? that overlayfs' .st_dev handling is broken? Do they still return the underlying dev_t? That's really really broken in overlayfs and will trip up everything else too, that relies on that for detecting file system boundaries... For example `find` and so on...

**Comment by xnox** [comment]
I will try to trace this more. And i wonder if .st_dev is consistent, but intermittent. E.g. when things change in workdir=, before appearing in the upperdir= during filesystem operations. Given the two directories are brand new, and in the upper-most level, it is reasonable to expect that it should all just work. I wonder if we should be treating .st_dev from upper/lower/work dirs as equivalent and the same.

**Comment by poettering** [comment]
overlayfs should really fix their .st_dev handling. As long as it doesn't do that it's a really broken file system and will break stuff all across the ecosystem, i.e. eveything that does "file system crossing" checks.

I mean, we could in theory use path_is_mount_point() to detect mount points while copying, but that code is black magic too with ultimate fallbacks to stat() based check if the file systems don't do name_to_handle_at() properly. And it's really slow if we'd have to run a bunch of syscalls for each and every single file we copy for the unlikely case we are looking at a mount point.

I mean, the overlayfs folks must have a strategy regarding "cp -x" (or "find -xdev"). The code we copy files with does the exact same stuff as "cp -x" in this regard. I mean, if they come up with a solution for "cp -x" we are happy to adopt it too. But otherwise not sure what we even *could* do...

**Comment by filbranden** [comment]
It looks like mounting the overlayfs using [`xino=on`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/filesystems/overlayfs.txt?id=16149013f8391d3bd7bcbcbe3bfd2fd06ba110e4#n41) would make it return the same `st_dev` for all files under it.

But that option was just introduced [a couple of months back](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/fs/overlayfs/super.c?id=795939a93e600587e52c34fe02402b27ddda6017), so not really sure it's safe to use yet... Looks like it was first included in 4.17-rc1, so available in the official 4.17 which was just released.

Cheers,
Filipe

**Comment by filbranden** [comment]
On the other hand `st_dev` should be [fine on directories](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/filesystems/overlayfs.txt?id=16149013f8391d3bd7bcbcbe3bfd2fd06ba110e4#n27), so maybe changing that logic to only check for it when traversing a directory would be ok?

This will copy more files than it did when bind-mounting files, but I'm not sure that's really a bug in itself...

**Comment by poettering** [comment]
Fix waiting in #9213

**Comment by poettering** [comment]
Closing as #9213 is merged now
